01-messaging.md

```diff
@@ -476,6 +476,75 @@ every message maximally).
 Finally, the usage of periodic `ping` messages serves to promote frequent key
 rotations as specified within [BOLT #8](08-transport.md).
 
+### Peer storage
+
+## The `peer_storage` and `your_peer_storage` Messages
+
+Nodes that advertise the `option_provide_storage` feature offer storing
+arbitrary data for their peers. The data stored must not exceed 65531 bytes,
+which lets it fit in lightning messages.
+
+Nodes that advertise the `option_want_storage` feature want to have their
+data stored by their peers that support `option_provide_storage`.
+
+Nodes can verify that their `option_provide_storage` peers correctly store
+their data at each reconnection, by comparing the contents of the
+retrieved data with the last one they sent. However, nodes should not expect
+their peers to always have their latest data available.
+
+Nodes ask their peers to store data using the `peer_storage` message & expect peers to return them latest data using `your_peer_storage` message:
+
+1. type: 7 (`peer_storage`)
+2. data:
+   * [`u16`: `length`]
+   * [`length*byte`:`blob`]
+
+
+1. type: 9 (`your_peer_storage`)
+2. data:
+   * [`u16`: `length`]
+   * [`length*byte`:`blob`]
+
+
+Requirements:
+
+The sender of `peer_storage`:
+  - MUST offer `option_want_storage`.
+  - MAY send `peer_storage` whenever necessary (usually, everytime their configuration changes)
+  - MUST limit its `blob` to 65531 bytes.
+  - MUST encrypt the data in a way that can be validated when received.
+
+
+The receiver of `peer_storage`:
+  - If it offered `option_provide_storage`:
+    - if it has an open channel with the sender:
+      - MUST store the message.
+    - MAY store anyway (e.g. might get paid for the service).
+
+- If it does store the message:
+  - MAY delay storage to ratelimit peer to no more than one update per second.
+  - MUST respond with `your_peer_storage` containing `blob` after it is stored.
```

Comment:
we can also ACK with a sha256 of the data provided.

Reply:
I agree with @TheBlueMatt here, our proposal for this feature (https://github.com/lightning/bolts/pull/881) does not have an ACK because I didn't find that useful.

Reply:
It serves as assurance to the sender, although not foolproof, that it won't be excessively wasteful. Additionally, updating peer_storage won't be a frequent occurrence.

Reply:
Although I agree that it's not necessary.

Reply:
> Additionally, updating peer_storage won't be a frequent occurrence.

That's not a good assumption at all! It will on the contrary be _very_ frequent. The way we use it today is to update the backup at every signed change in a channel, to ensure we're able to resume channels immediately after data loss.

Reply:
> Don't make me hash the data, either

Would you settle for md5? Or do we need to go all the way to CRC32? Like, on one hand, I understand a resistance to performing extra compute, but on the other hand...it's 65k of data. There are already provisions for rate limiting, so how much does this really matter?

If we want to drop the ACK, I can live with that, but I'm not sure I understand what is behind the strong resistance.

Reply:
just to clarify: we would keep the blob response on re-start right? need someway of requesting our back-up from the peer. 

Also - if we do only send this on re-establish, would it maybe be a good idea to potentially put it back in the actual re-establish message as a tlv instead of adding a whole new message?

Reply:
ah - actually.. just realised that then it would only work for peers we have channels with... which might not be a wanted restriction

Reply:
The sole reason for ACK was to assure the sender, is their any other efficient method to do this?

Reply:
But does it really assure the sender of anything meaningful? If used to store every channel state change (which is what we've been doing for years), we don't want to pause the channel's operation waiting for an ACK, it adds unnecessary complexity and risk of force-closing. This is mostly being used by nodes who will regularly disconnect (e.g. mobile phones), who have an opportunity of checking the back-up returned on every connection reestablishment, which is IMO enough assurance that the storing node behaves correctly?

Reply:
I'd like to remind us that we didn't *need* ACKs for every channel update either, and yet we are nudging ourselves towards a simplified update protocol. Granted, the purpose of this is far less consequential than ACK'ing the channel updates themselves, so we can afford to be like "🤷🏻‍♂️ who knows if they got it, we'll just assume so".

My take here is that at minimum it helps with debugging, and the cost of the ACK seems rather trivial so avoiding it seems silly. Yes, it doesn't provide much, we can always send back the packet, or the hash, or some ID, and then just yeet the data into oblivion, but does it really cost us much to just ACK it?

Reply:
Removed ACK as a requirement. Still mentioned in the rationale so that implementations can decide for themselves.